### PR TITLE
Crazy barcode scanning jQuery plugin. Janky, but only mildy so

### PR DESCRIPTION
**Pull request by RobRuana**

### CODE REVIEW

File: barcode/templates/barcode_client.html [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        // from the element used to initialize the barcodeField.
+        targetField: null,
+
+        // Control how the target field is blurred/focused.
+        blurOnKeys: [],
+        refocusAfterBlurOnKeys: true,  // Refocuses if blurred from blurOnKeys
+        focusOnError: true,  // Always focus, regardless of blurOnKeys
+        focusOnComplete: true,  // Always focus, regardless of blurOnKeys
+
+        // Turn off toastr error messages entirely.
```
Comment (kitsuta) [inline barcode/templates/barcode_client.html]: This is a confusing comment when paired with that name. 😅 Does this var SHOW error messages, as its name implies, or HIDE them, as the comment says? Same question for the variable after this one!

File: barcode/templates/barcode_client.html [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        showErrorMessage: true,
+
+        // Turn off toastr error messages, but only when target field is hidden.
+        showErrorMessageWhenHidden: true,
+
+        // Determine if we do any special data handling.
+        // detectBadgeNum and detectUUID are mututally exclusive. If both are
```
Comment (kitsuta) [inline barcode/templates/barcode_client.html]: Minor nitpick: it's easier to read comments if they don't split across lines like this. I THINK `// If both true, then only detectBadgeNum is used.` will fit on one line?

File: barcode/templates/barcode_client.html [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        detectUUID: false,
+
+        // Determine if the target field is auto-filled.
+        autoFill: true,
+
+        // Optional function to determine if target field should actually
+        // be auto-filled. Return true to auto-fill.
+        // Ignored if autoFill is falsey.
+        // function() { return true; }
+        shouldAutoFill: null,
```
Comment (kitsuta) [inline barcode/templates/barcode_client.html]: I was a little confused by this at first -- suggest renaming to `autoFillWhen` and the submit version to `autoSubmitWhen` to make their purpose a little more obvious.

In fact, if you do that, you can even rename `autoFill` to `shouldAutoFill` to differentiate it more from `autoSubmitForm`, which at first seems like its parallel when in fact they are different types of values. If... if that makes sense.

File: barcode/templates/barcode_client.html [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+  var scannerDetectionOnCompleteHandlers = [];
+  var scannerDetectionOnErrorHandlers = [];
+  var scannerDetectionOnReceiveHandlers = [];
+
+  $(function () {
+    $(document).scannerDetection({
+      timeBeforeScanTest: 100,
+      avgTimeByChar: 30,
+      endChar: [9, 13, '\\'],
+      startChar: [BADGE_NUM_CHAR, '\\'],
```
Comment (kitsuta) [inline barcode/templates/barcode_client.html]: I just realized, will this cause a problem for scanning UUIDs, which don't have the ~ character prepended like badge number barcodes do? If we have a scanner that doesn't add any characters, will it fail to detect a scan? 😰

Comment (RobRuana) [reply barcode/templates/barcode_client.html]: For the UUID fields, we are just treating the scanner like a keyboard. We're not doing the fancy thing where we hide the raw data from the textfield like we do with badge numbers. It still detects the scan using the timing based code (which I confirmed works on an ancient reg laptop).

**Pull request merged**